src/renderers/dom/shared/__tests__/DOMPropertyOperations-test.js

```diff
@@ -261,4 +261,20 @@ describe('DOMPropertyOperations', () => {
       );
     });
   });
+
+  describe('getValueForProperty', function() {
+    it('uses getValueForAttribute when given a custom attribute', function() {
+      var node = document.createElement('div');
+
+      node.setAttribute('data-test', 'test');
+
+      var value = DOMPropertyOperations.getValueForProperty(
```

Comment:
It's because of the branch that @nhunzaker linked in the OP. `getValueForProperty` isn't currently being called in `diffHydratedProperties` for custom attributes. Instead it's using `getValueForAttribute`.